### PR TITLE
adding grpc interface to network state

### PR DESCRIPTION
**Pull request by navillasa**
Adds gRPC server and client library for supporting BoltDB crud requests.
Example client server can be found in `examples/netstate-client`.

### CODE REVIEW

File: protos/netstate/netstate.proto [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+message FilePath {
+  string path = 1;
+  // smallValue is a value too small to be broken up and stored
+  // in different places
+  string smallValue = 2;
+}
+
+// ListRequest is a request message for the List rpc call
+message ListRequest {
+  string request = 1;
```
Comment (coyle) [inline protos/netstate/netstate.proto]: what is request supposed to be ?

File: protos/netstate/netstate.proto [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// Copyright (C) 2018 Storj Labs, Inc.
```
Comment (coyle) [inline protos/netstate/netstate.proto]: This looks really good!

File: protos/netstate/gen.go [hunk 3]
```diff
@@ -0,0 +1,3 @@
+package netstate
```
Comment (coyle) [inline protos/netstate/gen.go]: copyright?

File: pkg/netstate/service.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"google.golang.org/grpc"
+
+	"github.com/storj/protos/netstate"
+	"github.com/storj/storage/boltdb"
+)
+
+// NewServer creates a new NetState Server
+func NewServer(logger *zap.Logger, db *boltdb.Client) *grpc.Server {
+	grpcServer := grpc.NewServer()
+	netstate.RegisterNetStateServer(grpcServer, &NetState{
```
Comment (coyle) [inline pkg/netstate/service.go]: Awesome!

File: pkg/netstate/netstate.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2018 Storj Labs, Inc.
```
Comment (coyle) [inline pkg/netstate/netstate.go]: assuming these are all just stubs

File: cmd/netstate-grpc/server/main.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	initializeFlags()
+
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		grpclog.Fatalf("failed to listen: %v", err)
+	}
+	var opts []grpc.ServerOption
+	grpcServer := grpc.NewServer(opts...)
+	pb.RegisterNetStateServer(grpcServer, newServer())
+	grpcServer.Serve(lis)
```
Comment (coyle) [inline cmd/netstate-grpc/server/main.go]: You should listen for a close and shutdown gracefully

File: protos/netstate/netstate.proto (hunk 1 above)
Comment (navillasa) [reply protos/netstate/netstate.proto]: It's just a request from the client to list all of the file paths saved in the db. I wasn't planning on `List()` taking any arguments initially, but protoc didn't like that..

File: pkg/netstate/netstate.go (hunk 5 above)
Comment (navillasa) [reply pkg/netstate/netstate.go]: Yeah, soon to be unstubbed!

File: cmd/netstate-grpc/server/main.go (hunk 6 above)
Comment (navillasa) [reply cmd/netstate-grpc/server/main.go]: I'm going to delete/re-do this file but will note for the future!

File: protos/netstate/netstate.proto (hunk 1 above)
Comment (coyle) [reply protos/netstate/netstate.proto]: ah ok this can just be
```proto 
message ListRequest {}
```

you don't have to define anything inside the message

Comment (navillasa) [reply protos/netstate/netstate.proto]: Oh nice, thanks!

File: protos/netstate/netstate.proto (hunk 2 above)
Comment (navillasa) [reply protos/netstate/netstate.proto]: Thanks!

**Comment by navillasa** [comment]
At a weird spot where I think both the grpc server and client need access to the NetState struct which currently looks like:
```
type NetState struct {
  DB     Client
  logger *zap.Logger
}
```
So the NewServer function instantiates this in `storj/pkg/netstate/service.go`, but the NewClient function instantiates the client differently (i.e. not as an instance of NetState). I think I need to change this, because as it is, the client won't be able to access Get/Put/List as methods.

File: cmd/netstate-rpc/server/main.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	// then there will need to be error handling.
+	logger, _ := zap.NewDevelopment()
+	if prod {
+		logger, _ = zap.NewProduction()
+	}
+	defer logger.Sync()
+
+	bdb, err := boltdb.New(logger, dbPath)
+	if err != nil {
+		return
```
Comment (jtolds) [inline cmd/netstate-rpc/server/main.go]: logger.Fatal here?

File: examples/netstate-client/main.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	client := proto.NewNetStateClient(conn)
+
+	ctx := context.Background()
+
+	// Examples file paths to be saved
+	fp := proto.FilePath{
+		Path:       []byte("welcome/to/my/file/journey"),
+		SmallValue: []byte("granola"),
```
Comment (jtolds) [inline examples/netstate-client/main.go]: lol

File: protos/netstate/netstate.proto [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  // Get formats and hands off a file path to get a small value from boltdb
+  rpc Get(GetRequest) returns (GetResponse);
+  // List calls the bolt client's List function and returns all file paths
+  rpc List(ListRequest) returns (ListResponse);
+  // Delete formats and hands off a file path to delete from boltdb
+  rpc Delete(DeleteRequest) returns (DeleteResponse);
+}
+
+// FilePath is a request message for the Put rpc call
+message FilePath {
```
Comment (jtolds) [inline protos/netstate/netstate.proto]: honestly maybe this should just be called PutRequest to match with everything else?

File: protos/netstate/netstate.proto [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// GetRequest is a request message for the Get rpc call
+message GetRequest {
+  bytes path = 1;
+}
+
+// ListRequest is a request message for the List rpc call
+message ListRequest {
+  bytes bucket = 1;
```
Comment (jtolds) [inline protos/netstate/netstate.proto]: this is also going to need to take a path, and eventually a starting key and a limit

File: protos/netstate/netstate.proto [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// PutResponse is a response message for the Put rpc call
+message PutResponse {
+  string confirmation = 1;
```
Comment (jtolds) [inline protos/netstate/netstate.proto]: in this case, do we actually need a response proto at all? i assume grpc returns error codes somehow else

File: protos/netstate/netstate.proto [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// GetResponse is a response message for the Get rpc call
+message GetResponse {
+  bytes smallValue = 1;
```
Comment (jtolds) [inline protos/netstate/netstate.proto]: eventually this will be a Pointer!

**Comment by dylanlott** [comment]
ur welcome @navillasa

**Pull request merged**